### PR TITLE
EODHP-245-investigate-tools-for-running-system-and-smoke-tests

### PR DESCRIPTION
**Pull request by sgillies-tpzuk**
PR to investigate frameworks for system tests and smoke tests.

I investigated:
- Prow
- JenkinsX
- TestKube
- Argo Workflows
- Tekton

Selected Argo workflows as it seems to have features to accomplish everything we want, is free and open source, and is easiest to use.

The readme in the root summarises the investigation. Each item considered has its own directory with a readme roughly describing how I found working with each. For those items that seemed viable, I installed them and did some basic testing. For those that did not seem viable, I wrote a summary why in its readme.

The prow directory has a lot of manifests that required tailoring to get it to work. Please don't bother reviewing these manifests other than to get the feel that its installation was far more complex than the others.

### CODE REVIEW

File: tekton/README.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+   - Any image can be used, and command can be set inline or mounted via file/config
+
+   3/5
+
+2. How easy is it to execute individual or groups of tests?
+
+   - Additional manifest required to run Task
+   - Tests can be executed via Tekton CLI
+   - Multiple tests can be run with one CLI command
+   - Pipelines can cal multiple tasks and pass arguments between them
```
Comment (hcollingwood) [inline tekton/README.md]: Call?

File: README.md [hunk 2]
```diff
@@ -1 +1,52 @@
 # System Test Trade-off
+
+## Issue
+
+We wish to be able to run system tests. For example, such a test might create a workspace, link a git repo to it, run a workflow defined in the repo, check that catalogue entries produced by it appear in the catalogue and check the output data.
+
+We also wish to be able to run smoke tests. These are non-disruptive tests which can run on a production platform, for example after a new deployment. These may also involve running fast-running test workflows in a test workspace, adding catalogue entries, etc.
+
+We need a framework to run these in. We have previously used Testkube in EODHP-231: TestKube system environment test suites
```
Comment (hcollingwood) [inline README.md]: Consistency between TestKube and testkube throughout PR

File: README.md [hunk 3]
```diff
@@ -1 +1,52 @@
...
+DONE
```
Comment (hcollingwood) [inline README.md]: Remove word?

File: argo-workflows/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+kubectl -n argo port-forward service/argo-server 2746:2746
+```
+
+Navigate to https://localhost:2746
+
+Click + Submit New Workflow and then Edit using full workflow options.
+
+## Summary
+
+Argo Workflows seems like a very strong choice. It was very easy to set up and install and seems very light weight.
```
Comment (hcollingwood) [inline argo-workflows/README.md]:
```suggestion
Argo Workflows seems like a very strong choice. It was very easy to set up and install and seems very lightweight.
```

File: tekton/tests/hello-world.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+kind: Task
+metadata:
+  name: hello
+spec:
+  steps:
+    - name: echo
+      image: alpine
+      script: |
+        #!/bin/sh
+        echo "Hello World"
```
Comment (hcollingwood) [inline tekton/tests/hello-world.yaml]: Missing new line at end of file (it looks like the only one)

File: README.md [hunk 6]
```diff
@@ -1 +1,52 @@
...
+but 1) it has been difficult to create tests declaratively rather than having to run CLI tools or use a UI, and 2) there seem to be some recent licence changes and it’s not clear if it’ll be possible to use a dashboard UI.
```
Comment (hcollingwood) [inline README.md]: Continue previous line rather than starting new one?

**Review by hcollingwood — APPROVED** [review]

**Review by tjellicoe-tpzuk — APPROVED** [review]
Really thorough investigation, no comments

**Pull request merged**